sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,585 @@
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return new UnboundedSource.CheckpointMark() {
+        @Override
+        public void finalizeCheckpoint() throws IOException {
+          // nothing to do
+        }
+      };
+    }
+
+    @Override
+    public byte[] getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public UnboundedMqttSource getCurrentSource() {
+      return source;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write and send a message to a MQTT server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract MqttConnectionConfiguration mqttConnectionConfiguration();
+    @Nullable abstract int qos();
+    @Nullable abstract boolean retained();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setMqttConnectionConfiguration(MqttConnectionConfiguration configuration);
+      abstract Builder setQos(int qos);
+      abstract Builder setRetained(boolean retained);
+      abstract Write build();
+    }
+
+    /**
+     * Define MQTT connection configuration used to connect to the MQTT broker.
+     *
+     * @param configuration The {@link MqttConnectionConfiguration} instance.
+     * @return The {@link Write} {@link PTransform} with the corresponding connection configuration.
+     */
+    public Write withMqttConnectionConfiguration(MqttConnectionConfiguration configuration) {
+      return toBuilder().setMqttConnectionConfiguration(configuration).build();
+    }
+
+    /**
+     * Define the MQTT message quality of service.
+     *
+     * <ul>
+     * <li>Quality of Service 0 - indicates that a message should
+     * be delivered at most once (zero or one times). The message will not be persisted to disk,
+     * and will not be acknowledged across the network. This QoS is the fastest,
+     * but should only be used for messages which are not valuable - note that
+     * if the server cannot process the message (for example, there
+     * is an authorization problem), then an
+     * {@link MqttCallback#deliveryComplete(IMqttDeliveryToken)}.
+     * Also known as "fire and forget".</li>
+     *
+     * <li>Quality of Service 1 - indicates that a message should
+     * be delivered at least once (one or more times).  The message can only be delivered safely if
+     * it can be persisted, so the application must supply a means of
+     * persistence using {@code MqttConnectOptions}.
+     * If a persistence mechanism is not specified, the message will not be
+     * delivered in the event of a client failure.
+     * The message will be acknowledged across the network.
```

Comment:
What exactly does this mean: does it mean that writing a message will wait for a subscriber to acknowledge it? How long will it wait? What if there's currently no subscriber, or if the subscriber forgets to acknowledge? Do the answers to this depend on whether or not the message is persisted? (I suppose if it's persisted by the broker, then the writer doesn't need to wait for a subscriber ack because it's now the broker's responsibility to ensure it's delivered? But I'm not familiar enough with Mqtt)